src/Microsoft.ComponentDetection.Detectors/pip/IPyPiClient.cs

```diff
@@ -84,54 +92,88 @@ public PyPiClient(IEnvironmentVariableService environmentVariableService, ILogge
 
     public static HttpClient HttpClient { get; internal set; } = new HttpClient(HttpClientHandler);
 
-    public async Task<IList<PipDependencySpecification>> FetchPackageDependenciesAsync(string name, string version, PythonProjectRelease release)
+    private static async Task<HttpResponseMessage> SendGetRequestAsync(
+        Uri uri,
+        HttpCompletionOption completionOption = HttpCompletionOption.ResponseContentRead)
     {
-        var dependencies = new List<PipDependencySpecification>();
+        using var request = new HttpRequestMessage(HttpMethod.Get, uri);
+        request.Headers.UserAgent.Add(ProductValue);
+        request.Headers.UserAgent.Add(CommentValue);
+        var response = await HttpClient.SendAsync(request, completionOption);
+        return response;
+    }
 
+    public async Task<IList<PipDependencySpecification>> FetchPackageDependenciesAsync(string name, string version, PythonProjectRelease release)
+    {
         var uri = release.Url;
-        var response = await this.GetAndCachePyPiResponseAsync(uri);
+        var key = new CacheDependencyKey(name, version, uri);
 
-        if (!response.IsSuccessStatusCode)
+        return (await this.cachedDependencies.GetOrCreateAsync(key, FetchDependencies)).ToList();
+
+        async Task<List<PipDependencySpecification>> FetchDependencies(ICacheEntry cacheEntry)
         {
-            this.logger.LogWarning("Http GET at {ReleaseUrl} failed with status code {ResponseStatusCode}", release.Url, response.StatusCode);
-            return dependencies;
-        }
+            cacheEntry.SlidingExpiration = CacheSlidingExpiration;
+            cacheEntry.Size = 1;
+            using var response = await SendGetRequestAsync(uri, HttpCompletionOption.ResponseHeadersRead);
 
-        var package = new ZipArchive(await response.Content.ReadAsStreamAsync());
+            if (!response.IsSuccessStatusCode)
+            {
+                this.logger.LogWarning(
+                    "Http GET at {ReleaseUrl} failed with status code {ResponseStatusCode}",
+                    release.Url,
+                    response.StatusCode);
+                return new List<PipDependencySpecification>();
+            }
 
-        var entry = package.GetEntry($"{name.Replace('-', '_')}-{version}.dist-info/METADATA");
+            // Store the .whl file on the disk temporarily
+            var tempFilePath = Path.GetTempFileName();
+            await using (var fileStreamWrite =
```

Comment:
Is this needed? I know these are temp files, but depending on how big the file can be, couldn't writing to disk cause errors for systems that are particularly low on disk space?

Reply:
We encountered the opposite problem very often. We had enough disk space, but got OOM when buffering (and caching) large .whl files into memory in the HTTP `SendAsync` part. In addition, the files get copies to a new `MemoryStream` the `ZipArchive` constructor because the HTTP stream (in the current code) is not seekable, so if the .whl file is 600 MB in size it will increase the peak memory to 1.2GB.

Perhaps we can make this behavior configurable, so that consumers will be able to select whether to use disk storage for the downloads?

Reply:
Maybe we can start by adding some telemetry on the size of some wheels. I am a bit worried that we would eat up all the disk space downloading the wheels to disk, as technically we'd download them twice in a given CI build (`pip install` + running CD).

A slider or a configurable limit on either:

1. The size of the given wheel (only write wheels smaller than x bytes to disk)
2. Total disk space used downloading wheels

sounds like the best options.

Reply:
How about we have 2 options:
- Store wheels in memory (Default)
- Store wheels on the disk
Would that be fine?

It would leave current behavior unchanged, but give an option to use disk space instead of process memory for those who want it. A more fine-grained control like you suggested can be added at a later point if necessary.

We can set this with an environment variable (or maybe even an `Options<T>`)